rig-core/src/semantic_routing.rs

```diff
@@ -0,0 +1,89 @@
+use serde::{Deserialize, Serialize};
+
+use crate::vector_store::VectorStoreIndex;
+
+pub struct SemanticRouter<V> {
+    store: V,
+    threshold: f64,
+}
+
+impl<V> SemanticRouter<V> {
+    pub fn builder() -> SemanticRouterBuilder<V> {
+        SemanticRouterBuilder::new()
+    }
+}
+
+impl<V> SemanticRouter<V>
+where
+    V: VectorStoreIndex,
+{
+    pub async fn select_route(&self, query: &str) -> Option<String> {
+        let res = self.store.top_n(query, 1).await.ok()?;
+        let (score, _, SemanticRoute { tag }) = res.first()?;
+
+        println!("{tag}, {score}");
+
+        if *score < self.threshold {
+            return None;
+        }
+
+        Some(tag.to_owned())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SemanticRoute {
```

Comment:
I want to normalize adding some more default derives, atleast `Debug` (i've ran into situations in debugging where something was missing `Debug`.)

Atleast for new structs, we can go back and fill in some derives in a later PR.